[card-group]: Refactor to use subgrid in favor of sameHeight

m4olivei opened the pull request:
### Related Ticket(s)

Closes #11998

### Description

This is a refactoring of card group to make use of CSS subgrid instead of the Javascript `sameHeight` logic on all the pieces of the card group.

### Changelog

**Changed**

- Use CSS subgrid for layout out the card elements within a card group

<!-- React and Web Component deploy previews are enabled by default. -->
<!-- To enable additional available deploy previews, apply the following -->
<!-- labels for the corresponding package: -->
<!-- *** "test: e2e": Codesandbox examples and e2e integration tests -->
<!-- *** "package: services": Services -->
<!-- *** "package: utilities": Utilities -->
<!-- *** "RTL": React / Web Components (RTL) -->
<!-- *** "feature flag": React / Web Components (experimental) -->

andy-blum commented:
I've put together a couple demo pages showing the implications of this change. 
* [using sameHeight](https://card-group-sameheight--carbon-demos.netlify.app/)
* [using subgrid](https://card-group-subgrid--carbon-demos.netlify.app/)

## Card contents width dependent on body text

If we remove the card's body text something is getting squished. Seems like a rule inherited from `card` about `justify-content` causing this.

<img width="879" alt="Screenshot 2024-10-04 at 10 23 30 AM" src="https://github.com/user-attachments/assets/207f288b-bdbb-4945-9b67-333b8455c5eb">

## I don't know that we need to subgrid all the way down into the main slot

After fixing the card width issue above, it's easier to see that the tags will never be able to line up in this scenario due to the grid lines. This would also likely end up happening if we had multiple `<p>` tags in a card as well.

<img width="895" alt="Screenshot 2024-10-04 at 10 25 30 AM" src="https://github.com/user-attachments/assets/d9e782a5-3233-4223-9768-c5881bae1f19">

## We should make sure grid placement is more explicit

Right now all items just get placed into whatever grid cell is open next. If we are more explicit with our placements we'll get a better result

Before:
<img width="878" alt="Screenshot 2024-10-04 at 10 26 29 AM" src="https://github.com/user-attachments/assets/35012de7-8a77-4baa-b507-2279b3849dd8">

After:
<img width="869" alt="Screenshot 2024-10-04 at 10 26 35 AM" src="https://github.com/user-attachments/assets/f5396a17-bdd1-43c8-a9b6-b3772c867d96">

The easiest way to do this might be to wrap all our slots in additional divs:

Current:
```html
<div part="container" class="cds--tile cds--card cds--tile--clickable cds--card--link">
  <slot name="image"></slot>
  <div part="wrapper" class="cds--card__wrapper ">
    <div part="content" class="cds--card__content">
      <slot name="eyebrow"></slot> 
      <slot name="heading"></slot>    
      <div part="copy" class="cds--card__copy">
        <slot></slot>
      </div>
      <slot name="pictogram"></slot>
      <slot name="footer"></slot>
    </div>
  </div>
</div>
```

Suggested:
```html
<div part="container" class="cds--tile cds--card cds--tile--clickable cds--card--link">
  <div part="image-wrapper"><slot name="image"></slot></div>  
  <div part="wrapper" class="cds--card__wrapper ">
    <div part="content" class="cds--card__content">
      <div part="eyebrow-wrapper"><slot name="eyebrow"></slot></div> 
      <div part="heading-wrapper"><slot name="heading"></slot></div>    
      <div part="copy" class="cds--card__copy">
        <slot></slot>
      </div>
      <div part="pictogram-wrapper"><slot name="pictogram"></slot></div>
      <div part="footer-wrapper"><slot name="footer"></slot></div>
    </div>
  </div>
</div>
```

In this way, the interaction with the grid is limited to the card-group and the card-group-item. Styles on slotted content shouldn't break our grid's alignment. Additionally, we can apply the necessary spacing to the wrapper divs. With spacing applied to those internal elements we can also hide them with use of `:has()`

```scss
:host(c4d-card-group-item) {

  [part="heading-wrapper"] {
    margin-bottom: 64px // Required spacing below heading
  }
  
  :not(:has([slot="heading"])) [part="heading-wrapper"] {
    display: none;
  }

}
```

kennylam commented on packages/styles/scss/components/card-group/_card-group.scss:
```diff
@@ -56,10 +56,40 @@
       border: 0;
+      grid-row: span 10;
+      grid-template-rows: subgrid;
+      margin-block-end: 2rem;
```
```suggestion
      margin-block-end: $spacing-07;
```

kennylam commented on packages/styles/scss/components/card-group/_card-group.scss:
```diff
@@ -84,6 +114,23 @@
+
+    &[grid-mode='narrow'] {
+      .#{$prefix}--card {
+        margin-block-end: 1rem;
```
```suggestion
        margin-block-end: $spacing-05;
```

kennylam reviewed (approved):
This is great, thanks @m4olivei! Just some spots to use tokens instead of hard-coded units.

andy-blum commented:
Thanks for the updates, alignments are looking much better now.

The pictogram one still has some weirdness, though:
* Pictograms seem to disable eyebrows
* Pictograms seem to create their own footer cta icons
* The body text is double-wrapped in p tags

<img width="1029" alt="Screenshot 2024-10-08 at 8 54 32 AM" src="https://github.com/user-attachments/assets/b6f079c1-15de-4b5c-b360-f5e1b11adc63">

<img width="1221" alt="Screenshot 2024-10-08 at 8 55 59 AM" src="https://github.com/user-attachments/assets/a8cdca1b-37ea-4a01-85cb-f9315ef415d6">

m4olivei commented:
We should be looking _pretty_ good now. Waiting on the Storybook, but here is a rundown of the feedback so far.

> This is great, thanks @m4olivei! Just some spots to use tokens instead of hard-coded units.

All set there, thanks @kennylam. Fixed in https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/12063/commits/2e3e15b9bf03aa45fe62b482fba442a54724045f

> I've put together a couple demo pages showing the implications of this change.

These are great, thanks @andy-blum .

> If we remove the card's body text something is getting squished. Seems like a rule inherited from card about justify-content causing this.

Yep, fixed in https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/12063/commits/3ab3d81e1419f0501648c4d4b608171c78ec9a0d

> I don't know that we need to subgrid all the way down into the main slot

This whole thing threw me off at first too. Originally I was trying to declare a subgrid on a grandchild element to the grid container and expecting it to like get placed on the nearest grid ancestor. I realized pretty quickly that it doesn't work that way. The child, which was just a regular old grid item gets placed in the parent grid first, in the first row and column. Then the grandchild getting `grid-template-rows: subgrid` makes no sense, b/c its immediate parent is just a grid item of the grandparent ... not a grid container. As it turns out, you have to have each element in the tree except for the leaves be `display: grid`. And we have to start up at the element that wraps all of our items, then go all the way down. You can effectively take out some child elements using `display: contents`, but that's not always appropriate.

> We should make sure grid placement is more explicit

Resolved in https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/12063/commits/29f5b4f34421320fa02553b02e2dd30cfab365ef also in https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/12063/commits/b72b45c7d28cb658d67f1eb20405267fe27371b9 . The wrappers work well.

The only thing is, I made changes to the card markup, which is widely used, so we'll need to be careful to regression test.

> The pictogram one still has some weirdness, though:
> Pictograms seem to disable eyebrows

Fixed in https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/12063/commits/b72b45c7d28cb658d67f1eb20405267fe27371b9

> The pictogram one still has some weirdness, though:
> Pictograms seem to create their own footer cta icons

I haven't been able to reproduce this one.

> The pictogram one still has some weirdness, though:
> The body text is double-wrapped in p tags

This is also the case on `main`. There is [some code](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/main/packages/web-components/src/components/card/card.ts#L347-L353) that is responsible in `updated`. Haven't touched that here, so I want to say it's out of scope for this.

andy-blum commented:
Looks like the only remaining alignment issue is some extra margin thrown on `.cds--card__copy` when pictograms are involved.

<img width="1058" alt="Screenshot 2024-10-08 at 2 49 54 PM" src="https://github.com/user-attachments/assets/f6c5b045-185b-420a-b0c8-b921b303b088">

m4olivei commented:
> Looks like the only remaining alignment issue is some extra margin thrown on `.cds--card__copy` when pictograms are involved.

I'm not sure that's worth fixing TBH. Pictogram margins appear to be intentionally unique. I don't think it's worth trying to make them conform in the mixed card type case, when my read is that Pictograms are meant to appear in the card-group exclusively with other pictograms.

andy-blum commented:
We do need to make sure the spacing is correct when pictograms are used, however.

<img width="714" alt="Screenshot 2024-10-08 at 5 01 26 PM" src="https://github.com/user-attachments/assets/3a65d0c2-c15d-4780-8e55-02040423a28c">

<img width="368" alt="Screenshot 2024-10-08 at 5 01 41 PM" src="https://github.com/user-attachments/assets/d405fc99-816a-4780-8e19-a2b62aa50a23">

andy-blum reviewed (approved):
I think we're in a place where we can declare this good enough and an improvement over the current card-group setup that we should merge it in.
